crypto/rand/drbg_lib.c

```diff
@@ -410,34 +575,43 @@ static int drbg_bytes(unsigned char *out, int count)
     return ret;
 }
 
+/* Implements the default OpenSSL RAND_add() method */
 static int drbg_add(const void *buf, int num, double randomness)
 {
-    unsigned char *in = (unsigned char *)buf;
-    unsigned char *out, *end;
+    int ret = 0;
+    RAND_DRBG *drbg = RAND_DRBG_get0_global();
 
-    CRYPTO_THREAD_write_lock(rand_bytes.lock);
-    out = &rand_bytes.buff[rand_bytes.curr];
-    end = &rand_bytes.buff[rand_bytes.size];
+    if (drbg == NULL)
+        return 0;
 
-    /* Copy whatever fits into the end of the buffer. */
-    for ( ; --num >= 0 && out < end; rand_bytes.curr++)
-        *out++ = *in++;
+    if (num < 0 || randomness < 0.0)
+        return 0;
 
-    /* XOR any the leftover. */
-    while (num > 0) {
-        for (out = rand_bytes.buff; --num >= 0 && out < end; )
-            *out++ ^= *in++;
+    if (randomness > (double)drbg->max_entropylen) {
+        /*
+         * The purpose of this check is to bound |randomness| by a
+         * relatively small value in order to prevent an integer
+         * overflow when multiplying by 8 in the rand_drbg_restart()
+         * call below.
+         */
+        return 0;
     }
 
-    CRYPTO_THREAD_unlock(rand_bytes.lock);
-    return 1;
+    CRYPTO_THREAD_write_lock(drbg->lock);
+    ret = rand_drbg_restart(drbg, buf,
+                            (unsigned int)num, 8*(unsigned int)randomness);
```

Comment:
I'm wondering if we're actually able to restart it. If do_rand_drbg_init() failed, doesn't RAND_DRBG_get0_global() always return NULL?

Reply:
I changed the behaviour of `drbg_setup()` in order to make temporary failures of the entropy source recoverable: A failure of `RAND_DRBG_instantiate()` is now ignored in `drbg_setup()`, so the latter will fail only for some fatal reason. This was necessary, because as you noted, a failure of `do_rand_drbg_init()` would make the global DRBG inaccessible.